File: walk2.cpp
#include <functional>
#include <iostream>
#include <memory>
#include <optional>
#include <string>
#include <vector>

struct Node {
    std::string name;
    std::vector<std::shared_ptr<Node>> kids;
    std::weak_ptr<Node> parent;
};

auto init_parents(std::shared_ptr<Node> tree) -> void {
    for (auto& kid : tree->kids) {
        kid->parent = tree;
        init_parents(kid);
    }
}

auto walk(
    const Node& tree,
    std::function<void(const Node& node, int depth)> action,
    int depth = 0
) -> void {
    action(tree, depth);
    for (const auto& kid : tree.kids) {
        walk(*kid, action, depth + 1);
    }
}

auto print(const Node& tree) -> void {
    walk(tree, [](const auto& node, auto depth) {
        std::cout << std::string(2 * depth, ' ') << node.name << "\n";
    });
}

auto calc_total_depth(const Node& tree) -> int {
    auto total = 0;
    walk(tree, [&total](const auto& _, auto depth) {
        total += depth;
    });
    return total;
}

auto process(std::shared_ptr<Node> intro) -> std::shared_ptr<Node> {
    auto tree = std::make_shared<Node>(Node {"root", {
        intro,
        std::make_shared<Node>(Node {"one", {
            std::make_shared<Node>(Node {"two"}),
            std::make_shared<Node>(Node {"three"}),
        }}),
        std::make_shared<Node>(Node {"four"}),
    }});
    init_parents(tree);
    // auto internal_intro = tree->kids[0];
    // tree->kids.push_back(std::make_shared<Node>(Node {"outro"}));
    // std::cout << internal_intro->name << "\n";
    print(*tree);
    auto total_depth = 0;
    for (auto i = 0; i < 200'000; i += 1) {
        total_depth += calc_total_depth(*tree);
    }
    std::cout << "Total depth: " << total_depth << "\n";
    return tree;
}

auto main() -> int {
    auto intro = std::make_shared<Node>(Node {"intro"});
    auto tree = process(intro);
    std::cout << intro->parent.lock()->name << "\n";
}

File: Walk2.cs
using System.Linq;
using System.Runtime.InteropServices;

class Walker2 {
    record class Node(
        string Name,
        List<Node> Kids
    ) {
        public WeakReference<Node?> Parent = new WeakReference<Node?>(null);
        public Node(string name) : this(name, new()) { }
    }

    void InitParents(Node tree) {
        foreach (var kid in tree.Kids) {
            kid.Parent.SetTarget(tree);
            InitParents(kid);
        }
    }

    delegate void WalkAction(Node node, int depth);

    void Walk(Node tree, WalkAction action, int depth = 0) {
        action(tree, depth);
        for (var i = 0; i < tree.Kids.Count; i += 1) {
            var kid = tree.Kids[i];
            Walk(kid, action, depth + 1);
        }
    }

    void Print(Node tree) {
        Walk(tree, (node, depth) => {
            Console.WriteLine($"{"".PadLeft(2 * depth)}{node.Name}");
        });
    }

    int CalcTotalDepth(Node tree) {
        var total = 0;
        Walk(tree, (_, depth) => {
            total += depth;
        });
        return total;
    }

    void Process(Node intro) {
        var tree = new Node("root", new List<Node> {
            intro,
            new("one", new List<Node> {
                new("two"),
                new("three"),
            }),
            new("four"),
        });
        InitParents(tree);
        // Test pointer stability.
        // var nodes = CollectionsMarshal.AsSpan(tree.Kids);
        var internalIntro = tree.Kids[0];
        tree.Kids.Add(new("outro"));
        // tree.Kids.Clear();
        // Print(nodes[0]);
        Print(internalIntro);
        // Print tree and calculate.
        Print(tree);
        var totalDepth = 0;
        foreach (var _ in Enumerable.Range(0, 200_000)) {
            totalDepth += CalcTotalDepth(tree);
        }
        Console.WriteLine($"Total depth: {totalDepth}");
    }

    void Run() {
        var intro = new Node("intro");
        Process(intro);
        // System.GC.Collect();
        Node? root = null;
        intro.Parent.TryGetTarget(out root);
        Console.WriteLine(root.Name);
    }

    static void Main(string[] args) {
        new Walker2().Run();
    }
}
